### PR TITLE
New APIs for querying loaded libraries and allowing preference of custom library paths over system paths

### LINKED ISSUE #130: What is jnr.ffi.LibraryOption.LoadNow for?
What is `jnr.ffi.LibraryOption.LoadNow` for?
JavaDoc says: 

> Load the library into memory immediately, instead of lazily loading it

I searched for usages, but did not find any. Is it a deprecated option?

### LINKED ISSUE #239: Loading library with same name as system library will have an unexpected library loaded
### Scenario:

This pertains mainly to GNU/Linux.
I want to use a local version of a library with the file name `libfoo.so`. 
If my system has in its default lib paths (such as `/lib/x86_64-linux-gnu`) a library with the same name and a higher version (such as `libfoo.so.3`) then that library (in the system paths) will be loaded despite adding my own library in the preferred search path. My one wins only if my version is greather than or equal to the one found in the system. The system finding technically will see mine first (expected) but sees it as inferior to the system one.

### Expected Behavior:

My added search paths are ALWAYS preferred, even if a better version exists in the system paths. This could be because I want to use a non versioned library of a library that happens to be found on the system. For example if I actually want to use my own `libc.so`, which could be different from the system's `libc.so.6`, or even just my own older version for some reason like `libc.so.4`. Obviously this example is contrived and a little ridiculous, but it may be the case that a library is, unknowingly to the consumers, actually installed on the client's machine, and with a higher version than the one distributed which may even have no version, leading to confusing errors that are tricky to debug.

### Solution

The solution is relatively simple and is in `Platform.Linux.locateLibrary()` where towards the end of the function there is version checking. We could add a `LibraryOption` to give consumers preference for higher and/or system versions, if for example they're distributing some common library for compatibility but would prefer to use the system's one, which could be of a higher version.

I think a great API addition related to this, (and one I would personally find great use in) is some way to query which library am I actually using. So that I can be aware at runtime, the path of the library I am currently using.

I have begun to implement a solution to this on the [`javadocs` branch of my fork](https://github.com/basshelal/jnr-ffi/tree/javadocs) but it will need some deep internal changes.

Currently, libraries are loaded lazily, in this case meaning when there are declared functions in the mapping interface that need to be bound to native functions in the library. Only if a function exists in that interface will JNR-FFI actually attempt to load the library and look for the address of the function with that name. In the case of an empty interface it will silently fail (or succeed), this is *somewhat* fine because it doesn't actually matter, the interface was empty but when you fill the interface, *then* the loading happens and only then will we be able to check for success, and which paths were successful etc. Also how would we be able to query these successful paths, ie which place do consumers call, `Runtime` is a good candidate but implementing that could be difficult. Adding another method to `LoadedLibrary` is expensive since that's another special function name we need to be aware of. The actual successful paths are determined in `NativeLibrary` so we need it to be from something that can access that.

The ideal scenario is something like:

```java
LibC libc = LibraryLoader.loadLibrary(
            LibC.class,
            libraryOptions,
            searchPaths,
            libName
);

String libPath = Runtime.getLibraryPath(libc);

```

I'm going to try to do both of these myself in my fork. The main issue is easy enough but the important addition to the API is not (at least from what I can tell).

### PR DESCRIPTION
**Pull request by basshelal**
This PR adds the following features in order from most significant to least:

### `Runtime.getLoadedLibraries()`

Added a new static method to `Runtime` called `getLoadedLibraries()` which allows consumers to query all loaded libraries at any given point in time. This method returns a `List<NativeLibrary.LoadedLibraryData>` which is a simple data class that holds information about a loaded library such as names, search paths and successful paths. `getLoadedLibraries()` only returns *loaded* libraries, that is, libraries that have been loaded into memory successfully, ie a successful call to `dlopen()` was made. A library that has not yet loaded (because of lazy loading) will not appear, and a library that has been *unloaded* because it was GC'd will also not appear. This addresses an API request in #239 

### `LibraryOption.PreferCustomPaths` (GNU/Linux)

(Relevant to GNU/Linux only) Added new `LibraryOption.PreferCustomPaths` which tells JNR-FFI to prefer the libraries found in custom paths over those found in system paths, even if the system paths have a higher version. This option allows consistent behavior if you are distributing libraries since JNR-FFI will now always prefer your libraries over any (accidental or otherwise) system libraries with the same name or a higher version. For example, if you distribute a `libfoo.so` and the user's system paths *happen* to have a `libfoo.so.6`, if you use `LibraryOption.PreferCustomPaths`, JNR-FFI will prefer yours despite it having a lower version than the system's. This addresses #239 

### `LibraryOption.LoadNow` is now being used

`LibraryOption.LoadNow` will now actually attempt to load library ASAP, invalid libraries with empty interfaces will now fail if `LibraryOption.LoadNow` was added as a libraryOption in `LibraryLoader`. This option was unused previously as mentioned in #130 

### Misc

`UnsatisfiedLinkError` thrown when library was not found now contains details about the library names and search paths used to locate library, much more useful for debugging why a library was not found

This PR should fix #239 and fix #130 

I have also added thorough documentation and tests to support the new APIs and most of the code is commented for clarity.

### CODE REVIEW

**Comment by basshelal** [comment]
Ouch! Github actions check failed 😕, I think I know why but this Maven test runner is not very helpful in detailing which method fails. 

I have a suspicion which I will try right now.

**Comment by basshelal** [comment]
@headius Ready to merge. Take your time with this I'm in no hurry.

Out of curiosity, how do you debug failing tests in GitHub actions? The test runner we're using dumps reports but I don't know how one is able to access files from a GitHub actions server. Took me a while before I figured out what exactly was failing.

Do we need to change our test runner or some configuration to make the failures/reports show up in the console instead of in files? It was actually super frustrating and could be a little discouraging for any future contributors if their builds keep failing and they don't know why.

Or I've completely missed something super obvious, not unlikely 😸

**Comment by headius** [comment]
Wow, lots here to review. Thank you! I will have a look the next few days and get back to you if any changes seem necessary.

**Comment by headius** [comment]
@basshelal Can you clean up the commit history and get rid of the experimental commits? Otherwise I can squash merge as a single commit, but I would like to preserve the relevant parts of the history.

File: src/main/java/jnr/ffi/LibraryLoader.java
Comment (headius) [inline src/main/java/jnr/ffi/LibraryLoader.java:168]: Personally I think this form is fine but we should probably move the for loop within an `if {` block rather than having it inline.

File: src/main/java/jnr/ffi/LibraryLoader.java
Comment (headius) [inline src/main/java/jnr/ffi/LibraryLoader.java:478]: If we are going to `trim().isEmpty()` anyway we don't need the first check here.

File: src/main/java/jnr/ffi/provider/jffi/AsmLibraryLoader.java
Comment (headius) [inline src/main/java/jnr/ffi/provider/jffi/AsmLibraryLoader.java:35]: These imports appear to have been out of order anyway, but in the future try to do non-code changes (whitespace, import cleanup, etc) as separate commits.

**Review by headius — APPROVED** [review]
It looks ok to me, though I do not have the background to confirm that the search logic is appropriate for Linux (especially envs with custom/funky lib paths like Nix). I will request some reviews by others.

**Comment by headius** [comment]
Hoping to get some review from @fzakaria @enebo before we merge this but I am good with it.

**Comment by fzakaria** [comment]
Why is `LibraryOption.PreferCustomPaths` needed?
Isn't that the purpose of having the `java.library.path` be searched FIFO?

File: src/main/java/jnr/ffi/LibraryLoader.java
Comment (fzakaria) [reply src/main/java/jnr/ffi/LibraryLoader.java:478]: What's the purpose of this method?

File: src/main/java/jnr/ffi/LibraryOption.java
Comment (fzakaria) [inline src/main/java/jnr/ffi/LibraryOption.java:82]: Rather than `PreferCustomPaths`, maybe `RespectLibraryPath` sounds interesting.
Should the pre-existing option be a default and included here now: `PreferLatestVersion`

**Review by fzakaria — COMMENTED** [review]
(will continue reivewing.)
Happy to review. This might be a great opportunity to remove the default search paths also :D

Thank you for your contribution.

**Comment by basshelal** [comment]
> @basshelal Can you clean up the commit history and get rid of the experimental commits? Otherwise I can squash merge as a single commit, but I would like to preserve the relevant parts of the history.

@headius I'm not sure how I can do that, (still learning the hard parts of git) I don't mind if you squash it all into 1 commit, that actually makes sense to be honest.

File: src/main/java/jnr/ffi/LibraryLoader.java
Comment (basshelal) [reply src/main/java/jnr/ffi/LibraryLoader.java:478]: @headius Ah yes that's true, good eye!

@fzakaria If you mean the method `addPathsFromFile()` that was there before my changes, basically it adds all paths from a file, this is often `/etc/ld.so.conf` and/or the files located in `/etc/ld.so.conf.d/`.
In my Linux machine I would often get comments and empty lines, this is wasteful and can be ignored. There are `include` directives which I would like to add functionality for in the future since it can happen that a user/application may edit the master `/etc/ld.so.conf` file and add a new directory by doing `include $DIR` although this is rare as far as I'm aware

File: src/main/java/jnr/ffi/LibraryOption.java
Comment (basshelal) [reply src/main/java/jnr/ffi/LibraryOption.java:82]: @fzakaria I want to add as few options as possible, also we should keep the current behavior as the default for backwards compatibility even if I personally don't like it, I don't want upgrading users to get confused. Hence my decision for the new option to be the one that isn't the current default. What do you think about the name of the option @headius ?

**Comment by basshelal** [comment]
> Why is `LibraryOption.PreferCustomPaths` needed?
> Isn't that the purpose of having the `java.library.path` be searched FIFO?

@fzakaria I'm not sure I get what you mean.
My idea is to keep the current behavior as the default and allow callers to explicitly state they want their version (the one they provided a custom path for, lets say they built it and are packaging it with their application in a directory called `libs/`) even if there exists a version with a higher number on the system's defaults.

I experienced this first-hand with PortAudio, I wanted to use my own build from `libs/libportaudio.so` but JNR-FFI kept choosing a version that was on my system `/lib/x86_64-linux-gnu/libportaudio.so.2` (that I was unaware of by the way) because my version had no number, thus to force my version I had to rename mine to `libportaudio.so.99` to ensure I will always win the version comparison. This is ugly, and I only realized this after some time debugging and figuring out that my library was not the one being called.

Regarding removing the default search paths, I agree with that *but* this logic is used across all unixes, not just GNU/Linux. GNU/Linux is the only unix I know of that uses the `/etc/ld.so.conf` model, maybe modern BSDs also have it, but definitely not Darwin (macOS), I checked this myself on my macOS machine, hence we should still keep the legacy way of forcing default search paths like `/lib/`, `/usr/lib/` etc these are technically *"official"* Unix library paths anyway, so we're not breaking any rules by doing this manually.

Hope this clears up any confusions, and thanks for your kind words 😊

File: src/main/java/jnr/ffi/Platform.java
Comment (fzakaria) [inline src/main/java/jnr/ffi/Platform.java:568]: Not sure I get why the libraryPaths size is checked against the default size

File: src/main/java/jnr/ffi/Platform.java
Comment (fzakaria) [inline src/main/java/jnr/ffi/Platform.java:544]: Should this `Map` be a Set?

File: src/main/java/jnr/ffi/LibraryLoader.java
Comment (fzakaria) [inline src/main/java/jnr/ffi/LibraryLoader.java:483]: What is this change doing?

**Review by fzakaria — COMMENTED** [review]
I think it's a bit over engineered the "PreferCustomPaths" code in `locateLibrary`.
Here are my take-away and suggestions:

- Add the default as a LibraryOption `PreferHighestVersion`
- The original `locateLibrary(String, List<String>)` should call `locateLibrary(String, List<String>, Set<LibraryOption>)` with `PreferHighestVersion`
- The other `LibraryOption` should be `RespectLibraryPath` and not do anything fancy and simply iterate the list of library paths provided FIFO and return _the first match_
- Break up the pull-request into the individual additions so they are easier to review

**Comment by fzakaria** [comment]
Thanks for the clarification above; threw down some feedback but please defer to @headius for final-say on anything.
The PR overall was big; I think simplifying it into smaller reviews might be simpler (and help with the whole _git_ thing).
:+1:

File: src/main/java/jnr/ffi/Platform.java
Comment (basshelal) [reply src/main/java/jnr/ffi/Platform.java:544]: Yes but the rest of the APIs all use `Map<LibraryOption, Object>` when taking in the `options`, I'm simply following the standard. This makes things consistent and easy for any future changes too so that we don't have to convert from `Map` to `Set`, but yeah you are indeed right, `Set` definitely makes more sense overall.

File: src/main/java/jnr/ffi/Platform.java
Comment (basshelal) [reply src/main/java/jnr/ffi/Platform.java:568]: This is for the following edge case scenario that I want us to support:

Caller explicitly uses one of the default system paths as a custom path, for example they explicitly want `/lib/x86_64-linux-gnu/` instead of `/usr/lib/`. 

In this case the caller *prefers* a system path, our logic would fail for this scenario without this additional code, and would end up picking whichever one came first even though the caller **specifically requested** a certain path that *happened* to be a system one

I agree it's a little over-engineering, but it makes the behavior much more strictly defined, that is:

If you use `LibraryOption.PreferCustomPaths` your custom paths will ***always*** be preferred, no matter what or where they actually are. Otherwise, we will pick for you the highest version we find first.

File: src/main/java/jnr/ffi/LibraryLoader.java
Comment (basshelal) [reply src/main/java/jnr/ffi/LibraryLoader.java:483]: Ignoring empty lines, comments and `include` directives found in `*.so.conf` files. I would like to one day add support for these `include` directives.
These lines are wasteful and should be ignored

**Comment by basshelal** [comment]
> * Add the default as a LibraryOption `PreferHighestVersion`
> * The original `locateLibrary(String, List<String>)` should call `locateLibrary(String, List<String>, Set<LibraryOption>)` with `PreferHighestVersion`
> * The other `LibraryOption` should be `RespectLibraryPath` and not do anything fancy and simply iterate the list of library paths provided FIFO and return _the first match_

Hmm, I want to add as few `LibraryOption`s as possible, I also really don't want to add a new `LibraryOption` that just reflects the default behavior, that seems a little useless/wasteful. Could even be confusing for callers.

Let me see what @headius has to say about this.

> * Break up the pull-request into the individual additions so they are easier to review

I agree, I didn't think this through very well and I'll try to improve on that, this is my first major open source contribution so I hope you can forgive me for this and I hope I will improve on that front.

File: src/main/java/jnr/ffi/LibraryOption.java
Comment (headius) [reply src/main/java/jnr/ffi/LibraryOption.java:82]: I generally advocate that all possible options should be enumerated, so that it is possible to force the default behavior if for some reason that default is no longer used (embedded use, custom build, future behavior change).

**However** this is not the approach that has been taken thus far, for example the `LoadNow` option does not have an equivalent `LazyLoad` option. I think we can skip adding the default `PreferLatestVersion` option until we revisit how options should be used.

Comment (headius) [reply src/main/java/jnr/ffi/LibraryOption.java:82]: It is also trivial and non-breaking for us to add such an option in the future, if it turns out to be needed by some user.

Comment (basshelal) [reply src/main/java/jnr/ffi/LibraryOption.java:82]: @headius My exact thoughts, you said it better than I could! 😁

**Comment by headius** [comment]
I think we can go ahead and merge this. @fzakaria had some good input on style and design but most of it would not change any behavior. Additional cleanup can be done in other PRs and this one is already big.

I will leave this for one more day for additional input but then I vote that we merge and do any appropriate cleanup separately.

**Comment by headius** [comment]
Received approval to merge from @fzakaria and @enebo so let's go!

**Pull request merged**
